flexmeasures/data/migrations/versions/f0ee99278f6f_added_flexmodel_to_generic_asset_model.py

```diff
@@ -9,12 +9,85 @@
 from alembic import op
 import sqlalchemy as sa
 
+from collections import defaultdict
+
 # revision identifiers, used by Alembic.
 revision = "f0ee99278f6f"
 down_revision = "cb8df44ebda5"
 branch_labels = None
 depends_on = None
 
+flex_model_fields = dict(
+    min_soc_in_mwh="soc-min",
+)
+
+
+def group_sensors_by_field(sensors, conn, generic_asset_table):
+    """
+    This function groups sensors by flexmodel fields(using old the names) and checks for value mismatches.
+    """
+    field_specs = []
+
+    # conmstruct the field specifications
+    for old_field_name, new_field_name in flex_model_fields.items():
+        field_spec = dict(
+            new_field_name=new_field_name,
+            old_field_name=old_field_name,
+            grouped=defaultdict(list),
+        )
+        field_specs.append(field_spec)
+
+    # iterate over the sensors and group them by field
+    # and check for value mismatches
+    for sensor in sensors:
+        # fetch the generic asset
+        sensor_generic_asset = conn.execute(
+            sa.select(
+                generic_asset_table.c.id,
+                generic_asset_table.c.attributes,
+            ).where(generic_asset_table.c.id == sensor.generic_asset_id)
+        ).fetchone()
+
+        if sensor_generic_asset is None:
+            raise Exception(
+                f"Generic asset not found for sensor {sensor.id} with asset_id {sensor.generic_asset_id}"
+            )
+
+        sensor_attrs = sensor.attributes or {}
+        asset_attrs = sensor_generic_asset.attributes or {}
+
+        for field_spec in field_specs:
+            old_field_name = field_spec["old_field_name"]
+
+            # check if old_field_name exist on both attributes on sensor and asset
+            sensor_val = sensor_attrs.get(old_field_name)
+            asset_val = asset_attrs.get(old_field_name)
+
+            if sensor_val is not None and asset_val is not None:
+                if sensor_val != asset_val:
+                    raise Exception(
+                        f"Value mismatch for '{old_field_name}' in sensor {sensor.id}: sensor={sensor_val}, asset={asset_val}"
+                    )
+
+            # check if old_field_name exist on sensor attributes
+            if sensor_val is not None:
+                field_spec["grouped"][sensor.generic_asset_id].append(sensor)
```

Comment:
If you keep the Asset instead of the asset ID here, you can avoid querying for the Asset later on.

```suggestion
                field_spec["grouped"][sensor_generic_asset].append(sensor)
```

Reply:
oh, i never thought that could work as its an object 🤔, I will try this